lib/charms/mongodb/v1/shards_interface.py

```diff
@@ -541,28 +560,102 @@ def _handle_changed_secrets(self, event) -> None:
+            self.charm.tls.request_certificate(param=None, internal=False)
+        elif cluster_auth_keyfile and not cluster_auth_tls and not tls_integrated:
+            logger.info("Cluster implements internal membership auth via keyFile")
+
+        # Copy over keyfile regardless of whether the cluster uses TLS or or KeyFile for internal
+        # membership authentication. If TLS is disabled on the cluster this enables the cluster to
+        # have the correct cluster KeyFile readily available.
+        key_file_contents = self.database_requires.fetch_relation_field(
+            event.relation.id, KEYFILE_KEY
+        )
+        self.update_keyfile(key_file_contents=key_file_contents)
+
+    # Future PR - status updating for inconsistencies with TLS (i.e. shard has TLS but
+    # config-server does not and vice versa or CA-mismatch)
+
+    def get_cluster_passwords(
+        self, event: RelationChangedEvent
+    ) -> Tuple[Optional[str], Optional[str]]:
+        """Retrieves shared cluster passwords."""
+        operator_password = self.database_requires.fetch_relation_field(
+            event.relation.id, OPERATOR_PASSWORD_KEY
+        )
+        backup_password = self.database_requires.fetch_relation_field(
+            event.relation.id, BACKUP_PASSWORD_KEY
+        )
+        return (operator_password, backup_password)
+
+    def sync_cluster_passwords(
+        self, event: RelationChangedEvent, operator_password: str, backup_password: str
+    ) -> None:
+        """Updates shared cluster passwords."""
+        try:
+            self.update_password(
+                username=OperatorUser.get_username(), new_password=operator_password
+            )
+            self.update_password(BackupUser.get_username(), new_password=backup_password)
+        except RetryError:
+            self.charm.unit.status = BlockedStatus("Shard not added to config-server")
+            logger.error(
+                "Failed to sync cluster passwords from config-server to shard. Shard cannot be added to config-server, deferring event and retrying."
+            )
+            event.defer()
+
     def _on_relation_changed(self, event):
         """Retrieves secrets from config-server and updates them within the shard."""
+        # TODO Future PR include TLS sainity check in pass_hook_checks
         if not self.pass_hook_checks(event):
             logger.info("Skipping relation joined event: hook checks re not passed")
             return
 
         # if re-using an old shard, re-set drained flag.
         self.charm.unit_peer_data["drained"] = json.dumps(False)
 
-        # TODO: Future PR better status message behavior
-        self.charm.unit.status = MaintenanceStatus("Adding shard to config-server")
+        # relation-changed events can be used for other purposes (not only adding the shard), i.e.
+        # password rotation, secret rotation, mongos hosts rotation
+        if not self._is_mongos_reachable() or not self._is_added_to_cluster():
```

Comment:
It looks like if any mongos fail to happen, we notify the user that the charm `Adding shard to config-server`.
I wouldn't say that it is always true.
In any case, status handling up to Mehdi

Reply:
I agree - let me improve

Reply:
better?

Reply:
thanks